implement django tests with custom runner

eleanorjboyd opened the pull request:
This PR is a working demo of how use a custom runner for Django which defines `kwargs["resultclass"] = UnittestTestResult` that allows for the results to get processed and sent exactly like unittest. This also allows for django tests to be run with `python manage.py test` so that setup and teardown work. This is an alternative solution to https://github.com/microsoft/vscode-python/pull/21468 which allows the extension to run Django tests exactly like the command line.

This is still an experiment, much more analysis will have to be conducted into if this is a feasible solution.

mh-firouzjah commented:
> https://github.com/microsoft/vscode-python/blob/c59eb26f9e34b99965122a663326a691a2e2664a/pythonFiles/unittestadapter/django_runner.py#L61

This change will be permanent in the settings.py module, and on the other hand, I don't think it's a good idea for a plugin to allow itself to make such a change in one of the most important project files.
Additionally, even reading information from this module may challenge your plugin in terms of security.
- It's possible to use `--testrunner` flag for `manage.py test` to avoid applying such a change in settings.py module
  - `--testrunner` TESTRUNNER:
     Tells Django to use specified test runner class instead of the one specified by the TEST_RUNNER setting.

eleanorjboyd commented:
Aha! Fabulous, I must have missed the ` --testrunner ` flag. So you are saying if I use this flag then the security concerns would be reduced? Want to make sure I understand your concern correctly. Thanks!

mh-firouzjah commented on pythonFiles/unittestadapter/django_runner.py:
```diff
@@ -0,0 +1,47 @@
+        ]
+        print("Running Django run tests with command: ", command)
+        try:
+            subprocess.run(" ".join(command), shell=True, check=True)
```
it may require space in-between ' 
`' '.join...

eleanorjboyd replied:
a space as the join argument? I think I already put one in between the " but ill check

mh-firouzjah commented:
I will review these changes as soon as I can and I'll be in touch.
Thank you for considering my suggestions.

mh-firouzjah commented:
I was working on [this patch](https://github.com/microsoft/vscode-python/compare/main...mh-firouzjah:vscode-python:mh-firouzjah-patch-v2). but reviewing your code here changed my mind.

According to what I had there and what you brought here:

- The `DiscoverRunner` has a built in method [`build_suite`](https://github.com/django/django/blob/21382e2743d06efbf5623e7c9b6dccf2a325669b/django/test/runner.py#L897) which is used for `discover`ing tests.
- What I was working on, had some unnecessary steps which your code simply fixed those.
- You can use this method for the discovery part as well(a solution for what I mentioned in previous comment) maybe it requires a second subclass of `DiscoverRunner` which only executes `build_suite` but with same idea of overriding `get_test_runner_kwargs`.
- The idea of using environment variables to indicate the extension shall be ready for Django Tests(`django_test_enabled` and `django_test_enabled.lower() == "true"` and so on), could be replaced with a more harmonic way of the same procedure of having an option in preparing test environment for the first time - in which a drop-down menu asks users to select either of `pytest`/`unittest` and some other questions to make the extension ready to discover and run tests. - in this case adding an option for `django_tests` and asking some more questions to get raid of any ENV Variable, is my suggestion.

eleanorjboyd commented:
Thanks for your expertise! Ill look into the build_suite method for the discovery part as well!

In terms of the UI, we are planning on having a way similar to what it is like now to setup Django testing. We are moving to a new style of how testing args work as proposed / discussed here: https://github.com/microsoft/vscode-python/issues/21845. With this change we might be changing around our menu flow but I was thinking it would look something like this:

click `configure tests` ->  dropdown of either `pytest` or `unittest` (but unittest includes a descriptor like "use unittest to config django tests") ->  click on unittest  -> dropdown of either `basic unittest` or `Django tests`  ->  user selects `Django test` it would be configured in their settings as an environment variable that therefore the person could change later but the flow would be similar.


 See below for how the Django env var would then show up in our proposed config json design:

```
{
      "name": "Django run config", 
      "args": ["-v"],
      "env": {
	"PYTHONPATH": "${workspaceFolder}",
        "DJANGO_TESTING": "true",
	"DJANGO_MANAGE_PY_PATH": "/path/to/file",
      },
},
```

mh-firouzjah commented:
hi👋
I want to say, `self.setup_test_environment()` is also an important step. the code I wrote in order to make Django environment up and ready is somehow a version of this method, so consider look in to this one as well.
Thanks to your explanation I got the idea of how you're going to use GUI to let people set up for Django tests. 
Now it seems because surly there will be a `manage.py` module, the two steps of discovery and running can be done easier:
- for the `run` part I like what you've done, that's clean and concise
- for the discovery I imagine two different solutions: 
  - Either using the same idea of what I've done before to read the content of the `manage.py` module and then set the DJANGO_SETTINGS_MODULE environment variable and a call to django.setup() then the rest of the job will be done without any extra work and the default/current implementation of extension's `unittest` will discover Django tests well.
  - Or totally a new approach: Calling `DiscoverRunner.setup_test_environment()` then  `DiscoverRunner.build_suite` to set up Django environment up and discover Django tests. But this may require more coding and maintenance. I am not sure about the pros and cons of this approach.

eleanorjboyd commented:
Hello again! Your suggestion for setting the DJANGO_SETTINGS_MODULE and call django.setup() sounds good to me. As I am not very familiar with Django can you provide an example where Django setup is required for test discovery? I am still trying to wrap my head around what setup provides for the actions taken during discovery.

Secondly I updated the proposal to feature what we have discussed here. I would appreciate any comments you could add there on our discovery discussion as I did not include that just yet and wanted to instead continue our discussion on it either here or in that thread.

Finally, I wanted to give you a bit of a timeline on this specific project. Due to scheduling, including a personal vacation and then housekeeping work in December I will not be able to make progress on this for the reminder of the calendar year. Additionally, since I want to implement this using env vars, those are currently not configurable with testing as of the rewrite and will require [this proposal](https://github.com/microsoft/vscode-python/issues/21845) to be finalized and implemented. Once that is done we can then expose env vars for people to set and to enable this Django testing design. We are not sure on the timeline for the testing args proposal but it is a substantial change that will require input from many stakeholders and therefore do not think it would be done before March of next year. I cannot promise any dates but wanted to give you a heads up about the timeline and that this feature, as it is dependent on the proposal, will not be completed for some time. I am sorry about this as I know Django testing has been highly requested for a while but appreciate your patience in the next few months as we look to make architecture changes that will enable us to add Django support. Thanks

mh-firouzjah commented:
> where Django setup is required for test discovery?

[Calling django.setup() is required for “standalone” Django usage](https://docs.djangoproject.com/en/4.2/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage)
Also from the old discussion look at [this](https://github.com/microsoft/vscode-python/issues/73#issuecomment-1288115792) and more noticeable [one](https://github.com/microsoft/vscode-python/issues/73#issuecomment-1425313917) from [carltongibson](https://github.com/carltongibson).
And also ignoring to use `django.setup()` will result in `django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.` exception. 
And as @carltongibson explained [here](https://github.com/microsoft/vscode-python/issues/73#issuecomment-1451520441):
> the basics are setting the DJANGO_SETTINGS_MODULE environment variable, and then calling django.setup().

Since in the last shot, you fixed the test running problem by using a custom test-runner, I think what @carltongibson mentioned about taking care of database creation and tearing-down and so on, will be covered behind the since(well, it is basically what they did themselves: `manage.py test`).
While using `manage.py test` will free us of taking care of any of those steps in the `running`  process, but in the discovery process no `manage.py test` is used:
> just from a Django perspective, calling `django.setup()` is going to be needed. It requires an import path to a settings module, either via `DJANGO_SETTINGS_MODULE`, `env var` or a `--settings CLI flag` to `django-admin`. (The default `manage.py` file sets a default for `DJANGO_SETTINGS_MODULE` to the `settings file` created with the `startproject` command, so that's not a bad fallback value.)

eleanorjboyd commented:
makes sense! Thanks for your reply and therefore we will use your original recommendation around setup for discovery!

dark0dave commented:
Has there been any progress on this?

eleanorjboyd commented:
Hello! As an update I will be beginning development work on this next week! The previous blocker, custom args, has gotten delayed but we want to get working on this feature. If I end up needing to start a new PR I will attach that here. Thanks!

eleanorjboyd commented:
closing in favor of https://github.com/microsoft/vscode-python/pull/23935, please check this one out and give me your feedback! Thanks
